File: sdlf-ddk-lightweight/data_lake/foundations/foundations_stack.py
# Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License").
# You may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


import os
from pathlib import Path
from typing import Any, Dict, Tuple

import aws_cdk as cdk
import aws_cdk.aws_dynamodb as ddb
import aws_cdk.aws_iam as iam
import aws_cdk.aws_kms as kms
import aws_cdk.aws_lakeformation as lf
import aws_cdk.aws_lambda as lmbda
import aws_cdk.aws_s3 as s3
import aws_cdk.aws_ssm as ssm
from aws_cdk.aws_s3_deployment import (BucketDeployment, ServerSideEncryption,
                                       Source)
from aws_cdk.custom_resources import Provider
from aws_ddk_core.base import BaseStack
from aws_ddk_core.resources import KMSFactory, LambdaFactory, S3Factory
from constructs import Construct


class FoundationsStack(BaseStack):
    def __init__(
        self,
        scope: Construct,
        construct_id: str,
        environment_id: str,
        resource_prefix: str,
        app: str,
        org: str,
        runtime: lmbda.Runtime,
        **kwargs: Any,
    ) -> None:
        self._environment_id: str = environment_id
        self._resource_prefix = resource_prefix
        super().__init__(
            scope,
            construct_id,
            environment_id,
            stack_name=f"{self._resource_prefix}-FoundationsStack-{self._environment_id}",
            **kwargs
        )

        self._app = app
        self._org = org

        # creates DDB tables an registers them with SSM parameters
        self._object_metadata = self._create_octagon_ddb_table(
            name=f"octagon-ObjectMetadata-{self._environment_id}",
            ddb_props={"partition_key": ddb.Attribute(name="id", type=ddb.AttributeType.STRING)},
        )

        self._datasets = self._create_octagon_ddb_table(
            name=f"octagon-Datasets-{self._environment_id}",
            ddb_props={"partition_key": ddb.Attribute(name="name", type=ddb.AttributeType.STRING)},
        )

        self._pipelines = self._create_octagon_ddb_table(
            name=f"octagon-Pipelines-{self._environment_id}",
            ddb_props={"partition_key": ddb.Attribute(name="name", type=ddb.AttributeType.STRING)},
        )
        self._peh = self._create_octagon_ddb_table(
            name=f"octagon-PipelineExecutionHistory-{self._environment_id}",
            ddb_props={"partition_key": ddb.Attribute(name="id", type=ddb.AttributeType.STRING)},
        )

        # creates "register-provider" which inserts/updates dataset/pipelines into DDB tables.
        self._create_register(runtime)

        # creates encrypted buckets and registers them in lake formation
        self._lakeformation_bucket_registration_role = self._create_lakeformation_bucket_registration_role()
        self._raw_bucket, self._raw_bucket_key = self._create_bucket(name="raw")
        self._stage_bucket, self._stage_bucket_key = self._create_bucket(name="stage")
        self._analytics_bucket, self._analytics_bucket_key = self._create_bucket(name="analytics")
        self._artifacts_bucket, self._artifacts_bucket_key = self._create_bucket(name="artifacts")
        self._athena_bucket, self._athena_bucket_key = self._create_bucket(name="athena")

        # pushes scripts from data_lake/src/glue/ to S3 "artifacts" bucket.
        self._glue_role = self._create_sdlf_glue_artifacts()

        # creates lambda layer from code in src/layers/data_lake_library
        self._data_lake_library = self._create_data_lake_library_layer()

    def _create_octagon_ddb_table(self, name: str, ddb_props: Dict[str, Any]) -> ddb.Table:

        table_name = name.split("-")[1]

        # ddb kms key resource
        table_key: kms.IKey = KMSFactory.key(
            self,
            id=f"{name}-table-key",
            environment_id=self._environment_id,
            description=f"{self._resource_prefix} {name.title()} Table Key",
            alias=f"{self._resource_prefix}-{name}-ddb-table-key",
            enable_key_rotation=True,
            pending_window=cdk.Duration.days(30),
            removal_policy=cdk.RemovalPolicy.DESTROY,
        )

        # ddb resource
        table: ddb.Table = ddb.Table(
            self,
            f"{name}-table",
            table_name=name,
            encryption=ddb.TableEncryption.CUSTOMER_MANAGED,
            encryption_key=table_key,
            billing_mode=ddb.BillingMode.PAY_PER_REQUEST,
            removal_policy=cdk.RemovalPolicy.DESTROY,
            point_in_time_recovery=True,
            **ddb_props,
        )

        # SSM for ddb table name
        ssm.StringParameter(
            self,
            f"{name}-table-name-ssm",
            parameter_name=f"/SDLF/DynamoDB/{table_name}",
            string_value=name,
        )
        return table

    def _create_register(self, runtime: lmbda.Runtime) -> None:

        self._register_function: lmbda.IFunction = LambdaFactory.function(
            self,
            id="register-function",
            environment_id=self._environment_id,
            code=lmbda.Code.from_asset(os.path.join(f"{Path(__file__).parents[1]}", "src/lambdas/register")),
            handler="handler.on_event",
            memory_size=256,
            description="Registers Datasets, Pipelines and Stages into their respective DynamoDB tables",
            timeout=cdk.Duration.seconds(15 * 60),
            runtime=runtime,
            environment={
                "OCTAGON_DATASET_TABLE_NAME": self.datasets.table_name,
                "OCTAGON_PIPELINE_TABLE_NAME": self.pipelines.table_name
            }
        )
        self.datasets.grant_read_write_data(self._register_function)
        self.pipelines.grant_read_write_data(self._register_function)

        self._register_provider = Provider(
            self,
            "register-provider",
            on_event_handler=self._register_function,
        )

    def _create_lakeformation_bucket_registration_role(self) -> iam.Role:
        list_all_buckets = iam.PolicyStatement(
            effect=iam.Effect.ALLOW,
            actions=["s3:ListAllMyBuckets"],
            resources=["arn:aws:s3:::*"],
        )

        list_sdlf_buckets = iam.PolicyStatement(
            effect=iam.Effect.ALLOW,
            actions=["s3:ListBucket"],
            resources=[
                f"arn:aws:s3:::{self._resource_prefix}-{self._environment_id}-{cdk.Aws.REGION}-{cdk.Aws.ACCOUNT_ID}-*"
            ],
        )

        get_sdlf_objects = iam.PolicyStatement(
            effect=iam.Effect.ALLOW,
            actions=[
                "s3:GetObject",
                "s3:GetObjectAttributes",
                "s3:GetObjectTagging",
                "s3:GetObjectVersion",
                "s3:GetObjectVersionAttributes",
                "s3:GetObjectVersionTagging",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:PutObject"
            ],
            resources=[
                f"arn:aws:s3:::{self._resource_prefix}-{self._environment_id}-{cdk.Aws.REGION}-{cdk.Aws.ACCOUNT_ID}-*"
            ],
        )

        lakeformation_bucket_registration_role: iam.Role = iam.Role(
            self,
            "lakeformation-bucket-registration-role",
            assumed_by=iam.ServicePrincipal("lakeformation.amazonaws.com"),
            inline_policies={
                "LakeFormationDataAccessPolicyForS3": iam.PolicyDocument(
                    statements=[
                        list_all_buckets,
                        list_sdlf_buckets,
                        get_sdlf_objects,
                    ]
                )
            },
        )

        return lakeformation_bucket_registration_role

    def _create_bucket(self, name: str) -> Tuple[s3.IBucket, kms.IKey]:
        bucket_key: kms.IKey = KMSFactory.key(
            self,
            id=f"{self._resource_prefix}-{name}-bucket-key",
            environment_id=self._environment_id,
            description=f"{self._resource_prefix} {name.title()} Bucket Key",
            alias=f"{self._resource_prefix}-{name}-bucket-key",
            enable_key_rotation=True,
            pending_window=cdk.Duration.days(30),
            removal_policy=cdk.RemovalPolicy.DESTROY
        )

        ssm.StringParameter(
            self,
            f"{self._resource_prefix}-{name}-bucket-key-arn-ssm",
            parameter_name=f"/SDLF/KMS/{name.title()}BucketKeyArn",
            string_value=bucket_key.key_arn,
        )

        bucket: s3.IBucket = S3Factory.bucket(
            self,
            id=f"{self._resource_prefix}-{name}-bucket",
            environment_id=self._environment_id,
            bucket_name=f"{self._resource_prefix}-{self._environment_id}-{cdk.Aws.REGION}-{cdk.Aws.ACCOUNT_ID}-{name}",
            encryption=s3.BucketEncryption.KMS,
            encryption_key=bucket_key,
            access_control=s3.BucketAccessControl.BUCKET_OWNER_FULL_CONTROL,
            block_public_access=s3.BlockPublicAccess.BLOCK_ALL,
            removal_policy=cdk.RemovalPolicy.RETAIN,
            event_bridge_enabled=True
        )

        ssm.StringParameter(
            self,
            f"{self._resource_prefix}-{name}-bucket-name-ssm",
            parameter_name=f"/SDLF/S3/{name.title()}Bucket",
            string_value=f"{self._resource_prefix}-{self._environment_id}-{cdk.Aws.REGION}-{cdk.Aws.ACCOUNT_ID}-{name}",
        )

        lf.CfnResource(
            self,
            f"{self._resource_prefix}-{name}-bucket-lakeformation-registration",
            resource_arn=bucket.bucket_arn,
            use_service_linked_role=False,
            role_arn=self._lakeformation_bucket_registration_role.role_arn,
        )

        bucket_key.add_to_resource_policy(
            iam.PolicyStatement(
                effect=iam.Effect.ALLOW,
                actions=[
                    "kms:CreateGrant",
                    "kms:Decrypt",
                    "kms:DescribeKey",
                    "kms:Encrypt",
                    "kms:GenerateDataKey*",
                    "kms:ReEncrypt*",
                ],
                resources=["*"],
                principals=[self._lakeformation_bucket_registration_role],
            )
        )
        return bucket, bucket_key

    def _create_data_lake_library_layer(self) -> lmbda.LayerVersion:
        data_lake_library_layer = lmbda.LayerVersion(
            self,
            "data-lake-library-layer",
            layer_version_name="data-lake-library",
            code=lmbda.Code.from_asset(os.path.join(f"{Path(__file__).parents[1]}", "src/layers/data_lake_library")),
            compatible_runtimes=[lmbda.Runtime.PYTHON_3_9],
            description=f"{self._resource_prefix} Data Lake Library",
            license="Apache-2.0",
        )
        return data_lake_library_layer

    def _create_sdlf_glue_artifacts(self) -> iam.IRole:

        bucket_deployment_role: iam.IRole = iam.Role(  # type: ignore
            self,
            f"{self._resource_prefix}-glue-script-s3-deployment-role",
            assumed_by=iam.ServicePrincipal("lambda.amazonaws.com"),
            managed_policies=[
                iam.ManagedPolicy.from_aws_managed_policy_name("service-role/AWSLambdaBasicExecutionRole")
            ],
        )

        self._artifacts_bucket_key.grant_encrypt_decrypt(bucket_deployment_role)

        glue_path = "data_lake/src/glue/"

        BucketDeployment(
            self,
            f"{self._resource_prefix}-glue-script-s3-deployment",
            sources=[Source.asset(f"{glue_path}")],
            destination_bucket=self._artifacts_bucket,
            destination_key_prefix=f"{glue_path}",
            server_side_encryption_aws_kms_key_id=self._artifacts_bucket_key.key_id,
            server_side_encryption=ServerSideEncryption.AWS_KMS,
            role=bucket_deployment_role,
        )

        glue_role: iam.IRole = iam.Role(  # type: ignore
            self,
            f"{self._resource_prefix}-glue-stageb-job-role",
            assumed_by=iam.ServicePrincipal("glue.amazonaws.com"),
            managed_policies=[iam.ManagedPolicy.from_aws_managed_policy_name("service-role/AWSGlueServiceRole")],
        )

        use_kms_key = iam.PolicyStatement(
            effect=iam.Effect.ALLOW,
            actions=[
                "kms:CreateGrant",
                "kms:Decrypt",
                "kms:DescribeKey",
                "kms:Encrypt",
                "kms:GenerateDataKey",
                "kms:GenerateDataKeyPair",
                "kms:GenerateDataKeyPairWithoutPlaintext",
                "kms:GenerateDataKeyWithoutPlaintext",
                "kms:ReEncryptTo",
                "kms:ReEncryptFrom"
            ],
            resources=[f"arn:aws:kms:{cdk.Aws.REGION}:{cdk.Aws.ACCOUNT_ID}:key/*"],
            conditions={"ForAnyValue:StringLike": {"kms:ResourceAliases": f"alias/{self._resource_prefix}-*-key"}},
        )

        list_sdlf_buckets = iam.PolicyStatement(
            effect=iam.Effect.ALLOW,
            actions=["s3:ListBucket"],
            resources=[
                f"arn:aws:s3:::{self._resource_prefix}-{self._environment_id}-{cdk.Aws.REGION}-{cdk.Aws.ACCOUNT_ID}-*"
            ],
        )

        get_sdlf_objects = iam.PolicyStatement(
            effect=iam.Effect.ALLOW,
            actions=[
                "s3:GetObject",
                "s3:PutObject",
                "s3:DeleteObject",
            ],
            resources=[
                f"arn:aws:s3:::{self._resource_prefix}-{self._environment_id}-{cdk.Aws.REGION}-{cdk.Aws.ACCOUNT_ID}-*"
            ],
        )

        access_glue_and_lakeformation = iam.PolicyStatement(
            effect=iam.Effect.ALLOW,
            actions=[
                "lakeformation:DeregisterResource",
                "lakeformation:GetDataAccess",
                "lakeformation:GrantPermissions",
                "lakeformation:PutDataLakeSettings",
                "lakeformation:GetDataLakeSettings",
                "lakeformation:RegisterResource",
                "lakeformation:RevokePermissions",
                "lakeformation:UpdateResource",
                "glue:CreateDatabase",
                "glue:CreateJob",
                "glue:CreateSecurityConfiguration",
                "glue:DeleteDatabase",
                "glue:DeleteJob",
                "glue:DeleteSecurityConfiguration",
                "glue:GetDatabase",
                "glue:GetDatabases",
                "glue:GetMapping",
                "glue:GetPartition",
                "glue:GetPartitions",
                "glue:GetPartitionIndexes",
                "glue:GetSchema",
                "glue:GetSchemaByDefinition",
                "glue:GetSchemaVersion",
                "glue:GetSchemaVersionsDiff",
                "glue:GetTable",
                "glue:GetTables",
                "glue:GetTableVersion",
                "glue:GetTableVersions",
                "glue:GetTags",
                "glue:PutDataCatalogEncryptionSettings",
                "glue:SearchTables",
                "glue:TagResource",
                "glue:UntagResource",
                "glue:UpdateDatabase",
                "glue:UpdateJob",
                "glue:ListSchemas",
                "glue:ListSchemaVersions"
            ],
            resources=["*"],
        )

        get_octagon_config = iam.PolicyStatement(
            effect=iam.Effect.ALLOW,
            actions=["dynamodb:GetItem"],
            resources=[
                f"arn:aws:dynamodb:{cdk.Aws.REGION}:{cdk.Aws.ACCOUNT_ID}:table/"
                + f"{self._resource_prefix}-{self._environment_id}-*",
                f"arn:aws:dynamodb:{cdk.Aws.REGION}:{cdk.Aws.ACCOUNT_ID}:table/octagon-*"
            ],
        )

        iam.ManagedPolicy(
            self,
            f"{self._resource_prefix}-glue-job-policy",
            roles=[glue_role],
            document=iam.PolicyDocument(
                statements=[
                    use_kms_key,
                    list_sdlf_buckets,
                    get_sdlf_objects,
                    access_glue_and_lakeformation,
                    get_octagon_config,
                ]
            ),
        )
        return glue_role

    @property
    def raw_bucket(self) -> s3.IBucket:
        return self._raw_bucket

    @property
    def raw_bucket_key(self) -> kms.IKey:
        return self._raw_bucket_key

    @property
    def stage_bucket(self) -> s3.IBucket:
        return self._stage_bucket

    @property
    def stage_bucket_key(self) -> kms.IKey:
        return self._stage_bucket_key

    @property
    def artifacts_bucket(self) -> s3.IBucket:
        return self._artifacts_bucket

    @property
    def artifacts_bucket_key(self) -> kms.IKey:
        return self._artifacts_bucket_key

    @property
    def glue_role(self) -> iam.IRole:
        return self._glue_role

    @property
    def register_provider(self) -> Provider:
        return self._register_provider

    @property
    def data_lake_library(self) -> lmbda.ILayerVersion:
        return self._data_lake_library

    @property
    def object_metadata(self) -> ddb.Table:
        return self._object_metadata

    @property
    def datasets(self) -> ddb.Table:
        return self._datasets

    @property
    def pipelines(self) -> ddb.Table:
        return self._pipelines

    @property
    def peh(self) -> ddb.Table:
        return self._peh

File: sdlf-ddk-lightweight/typescript/src/datalake/foundations/foundations-stack.ts
import * as path from "path";
import * as cdk from "aws-cdk-lib";
import * as dynamo from "aws-cdk-lib/aws-dynamodb";
import * as kms from "aws-cdk-lib/aws-kms";
import * as lakeformation from "aws-cdk-lib/aws-lakeformation";
import * as iam from "aws-cdk-lib/aws-iam";
import * as lambda from "aws-cdk-lib/aws-lambda";
import * as s3 from "aws-cdk-lib/aws-s3";
import * as ssm from "aws-cdk-lib/aws-ssm";
import * as s3Deployment from "aws-cdk-lib/aws-s3-deployment";
import * as cr from "aws-cdk-lib/custom-resources";
import { BaseStack, LambdaDefaults, S3Defaults } from "aws-ddk-core";

import { Construct } from "constructs";

export interface FoundationsStackProps extends cdk.StackProps {
  readonly environmentId: string;
  readonly resourcePrefix: string;
  readonly app: string;
  readonly org: string;
  readonly runtime: lambda.Runtime;
}
interface createOctagonTableProps {
  readonly name: string;
  readonly partitionKey: any;
}
export class FoundationsStack extends BaseStack {
  readonly resourcePrefix: string;
  readonly environmentId: string;
  readonly app: string;
  readonly org: string;
  readonly objectMetadata: dynamo.Table;
  readonly datasets: dynamo.Table;
  readonly pipelines: dynamo.Table;
  readonly peh: dynamo.Table;
  readonly registerProvider: cr.Provider;
  readonly lakeformationBucketRegistrationRole: iam.Role;
  readonly rawBucket: s3.Bucket;
  readonly rawBucketKey: kms.Key;
  readonly stageBucket: s3.Bucket;
  readonly stageBucketKey: kms.Key;
  readonly analyticsBucket: s3.Bucket;
  readonly analyticsBucketKey: kms.Key;
  readonly artifactsBucket: s3.Bucket;
  readonly artifactsBucketKey: kms.Key;
  readonly athenaBucket: s3.Bucket;
  readonly athenaBucketKey: kms.Key;
  readonly glueRole: iam.IRole;

  constructor(scope: Construct, id: string, props: FoundationsStackProps) {
    super(scope, `${props.resourcePrefix}-FoundationsStack-${props.environmentId}`, props);
    this.resourcePrefix = props.resourcePrefix
    this.environmentId = props.environmentId
    this.app = props.app;
    this.org = props.org;
    this.objectMetadata = this.createOctagonTable({
      name: `octagon-ObjectMetadata-${props.environmentId}`,
      partitionKey: {name: "id", type: dynamo.AttributeType.STRING},
    })

    this.datasets = this.createOctagonTable({
      name: `octagon-Datasets-${props.environmentId}`,
      partitionKey: {"partition_key": {name: "name", type: dynamo.AttributeType.STRING}},
    })
    this.pipelines = this.createOctagonTable({
      name: `octagon-Pipelines-${props.environmentId}`,
      partitionKey: {"partition_key": {name: "name", type: dynamo.AttributeType.STRING}},
    })
    this.peh = this.createOctagonTable({
      name: `octagon-PipelineExecutionHistory-${props.environmentId}`,
      partitionKey: {"partition_key": {name: "id", type: dynamo.AttributeType.STRING}},
    })
    this.createRegister(props.runtime)

    // creates encrypted buckets and registers them in lake formation
    this.lakeformationBucketRegistrationRole = this.createLakeformationBucketRegistrationRole();
    const [rawBucket, rawBucketKey] = this.createBucket("raw")
    this.rawBucket = rawBucket;
    this.rawBucketKey = rawBucketKey;
    const [stageBucket, stageBucketKey] = this.createBucket("stage")
    this.stageBucket = stageBucket;
    this.stageBucketKey = stageBucketKey;
    const [analyticsBucket, analyticsBucketKey] = this.createBucket("analytics")
    this.analyticsBucket = analyticsBucket;
    this.analyticsBucketKey = analyticsBucketKey;
    const [artifactsBucket, artifactsBucketKey] = this.createBucket("artifacts")
    this.artifactsBucket = artifactsBucket;
    this.artifactsBucketKey = artifactsBucketKey;
    const [athenaBucket, athenaBucketKey] = this.createBucket("athena")
    this.athenaBucket = athenaBucket;
    this.athenaBucketKey = athenaBucketKey;

    // pushes scripts from data_lake/src/glue/ to S3 "artifacts" bucket.
    this.glueRole = this.createSdlfGlueArtifacts()


  }
  private createOctagonTable(props: createOctagonTableProps): dynamo.Table {

    const tableName = props.name.split("-")[1]

    const tableKey = new kms.Key(
      this,
      `${props.name}-table-key`,
      {
        description: `${this.resourcePrefix} ${props.name} Table Key`,
        alias: `${this.resourcePrefix} ${props.name} Table Key`,
        enableKeyRotation: true,
        pendingWindow: cdk.Duration.days(30),
        removalPolicy: cdk.RemovalPolicy.DESTROY
      }
    )
    
    new ssm.StringParameter(
      this,
      `${props.name}-table-name-ssm`,
      {
        parameterName: `/SDLF/DynamoDB/${tableName}`,
        stringValue: props.name,
      }
    )
    return new dynamo.Table(
      this,
      `${props.name}-table`,
      {
        partitionKey: props.partitionKey,
        tableName: props.name,
        encryption: dynamo.TableEncryption.CUSTOMER_MANAGED,
        encryptionKey: tableKey,
        billingMode: dynamo.BillingMode.PAY_PER_REQUEST,
        removalPolicy: cdk.RemovalPolicy.DESTROY,
        pointInTimeRecovery: true,
      }
    )
  }
  private createRegister(runtime: lambda.Runtime): void {
    const registerFunction = new lambda.Function(
      this,
      "register-function",
      LambdaDefaults.functionProps(
        {
          code: lambda.Code.fromAsset(path.join(__dirname, "src/lambdas/register/")),
          handler: "handler.on_event",
          memorySize: 256,
          description: "Registers Datasets, Pipelines and Stages into their respective DynamoDB tables",
          timeout: cdk.Duration.minutes(15),
          runtime: runtime,
          environment: {
             "OCTAGON_DATASET_TABLE_NAME": this.datasets.tableName,
             "OCTAGON_PIPELINE_TABLE_NAME": this.pipelines.tableName
          }
        }
      ),  
    )
    this.datasets.grantReadWriteData(registerFunction)
    this.pipelines.grantReadWriteData(registerFunction)
    const registerProvider = new cr.Provider(
      this,
      "register-provider",
      {
        onEventHandler: registerFunction
      }
    )


  }
  private createBucket(name: string): [s3.Bucket, kms.Key] {
    const bucketKey = new kms.Key(
      this,
      `${this.resourcePrefix}-$${name}-bucket-key`,
      {
        description: `${this.resourcePrefix} $${name} Bucket Key`,
        alias: `${this.resourcePrefix}-$${name}-bucket-key`,
        enableKeyRotation: true,
        pendingWindow: cdk.Duration.days(30),
        removalPolicy: cdk.RemovalPolicy.DESTROY
      }
    )
    new ssm.StringParameter(
      this,
      `${this.resourcePrefix}-$${name}-bucket-key-arn-ssm`,
      {
        parameterName: `/SDLF/KMS/$${name}BucketKeyArn`,
        stringValue: bucketKey.keyArn,
      }
    )

    const bucket = new s3.Bucket(
        this,
        `${this.resourcePrefix}-$${name}-bucket`,
        S3Defaults.bucketProps(
          {
            bucketName: `${this.resourcePrefix}-${this.environmentId}-${cdk.Aws.REGION}-${cdk.Aws.ACCOUNT_ID}-$${name}`,
            encryption: s3.BucketEncryption.KMS,
            encryptionKey: bucketKey,
            accessControl: s3.BucketAccessControl.BUCKET_OWNER_FULL_CONTROL,
            blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
            removalPolicy: cdk.RemovalPolicy.RETAIN,
            eventBridgeEnabled: true
          }
        )
    )

    new ssm.StringParameter(
      this,
      `${this.resourcePrefix}-${name}-bucket-name-ssm`,
      {
        parameterName: `/SDLF/S3/${name}Bucket`,
        stringValue: `${this.resourcePrefix}-${this.environmentId}-${cdk.Aws.REGION}-${cdk.Aws.ACCOUNT_ID}-${name}`,
      }
    )

    new lakeformation.CfnResource(
      this,
      `${this.resourcePrefix}-${name}-bucket-lakeformation-registration`,
      {
        resourceArn: bucket.bucketArn,
        useServiceLinkedRole: false,
        roleArn: this.lakeformationBucketRegistrationRole.roleArn,
      }
    )

    bucketKey.addToResourcePolicy(
      new iam.PolicyStatement(
        {
          effect: iam.Effect.ALLOW,
          actions: [
            "kms:CreateGrant",
            "kms:Decrypt",
            "kms:DescribeKey",
            "kms:Encrypt",
            "kms:GenerateDataKey*",
            "kms:ReEncrypt*",
          ],
          resources: ["*"],
          principals: [this.lakeformationBucketRegistrationRole],
        }
      )
    )
    return [bucket, bucketKey]
  }
  private createLakeformationBucketRegistrationRole(): iam.Role {
    return new iam.Role(
      this,
      "lakeformation-bucket-registration-role",
      {
        assumedBy: new iam.ServicePrincipal("lakeformation.amazonaws.com"),
        inlinePolicies: {
          "LakeFormationDataAccessPolicyForS3": new iam.PolicyDocument(
            {
              statements: [
                new iam.PolicyStatement({
                  effect: iam.Effect.ALLOW,
                  actions: [
                      "s3:GetObject",
                      "s3:GetObjectAttributes",
                      "s3:GetObjectTagging",
                      "s3:GetObjectVersion",
                      "s3:GetObjectVersionAttributes",
                      "s3:GetObjectVersionTagging",
                      "s3:PutObjectTagging",
                      "s3:PutObjectVersionTagging",
                      "s3:PutObject"
                  ],
                  resources:[
                    `arn:aws:s3:::${this.resourcePrefix}-${this.environmentId}-${cdk.Aws.REGION}-${cdk.Aws.ACCOUNT_ID}-*`
                  ]
                }),
                new iam.PolicyStatement({
                  effect: iam.Effect.ALLOW,
                  actions: ["s3:ListBucket"],
                  resources: [
                    `arn:aws:s3:::${this.resourcePrefix}-${this.environmentId}-${cdk.Aws.REGION}-${cdk.Aws.ACCOUNT_ID}-*`
                  ],
                }),
                new iam.PolicyStatement({
                  effect: iam.Effect.ALLOW,
                  actions: ["s3:ListAllMyBuckets"],
                  resources: ["arn:aws:s3:::*"],
                })
              ]
            }
          )
        }
      }
    )
  }
  protected createSdlfGlueArtifacts(): iam.IRole {
    const bucketDeploymentRole = new iam.Role( 
      this,
      `${this.resourcePrefix}-glue-script-s3-deployment-role`,
      {
        assumedBy: new iam.ServicePrincipal("lambda.amazonaws.com"),
        managedPolicies: [
          iam.ManagedPolicy.fromAwsManagedPolicyName("service-role/AWSLambdaBasicExecutionRole")
        ],
      }
    )
    this.artifactsBucketKey.grantEncryptDecrypt(bucketDeploymentRole)
    const gluePath = "data_lake/src/glue/";

    new s3Deployment.BucketDeployment(
      this,
      `${this.resourcePrefix}-glue-script-s3-deployment`,
      {
        sources: [s3Deployment.Source.asset(gluePath)],
        destinationBucket: this.artifactsBucket,
        destinationKeyPrefix: gluePath,
        serverSideEncryptionAwsKmsKeyId: this.artifactsBucketKey.keyId,
        serverSideEncryption: s3Deployment.ServerSideEncryption.AWS_KMS,
        role: bucketDeploymentRole,
      }
    )

    const glueRole= new iam.Role(
      this,
      `${this.resourcePrefix}-glue-stageb-job-role`,
      {
        assumedBy: new iam.ServicePrincipal("glue.amazonaws.com"),
        managedPolicies: [iam.ManagedPolicy.fromAwsManagedPolicyName("service-role/AWSGlueServiceRole")],
      }
    )
    
    new iam.ManagedPolicy(
      this,
      `${this.resourcePrefix}-glue-job-policy`,
      {
        roles: [glueRole],
        document: new iam.PolicyDocument({
          statements: [
            new iam.PolicyStatement(
              {
                effect: iam.Effect.ALLOW,
                actions: [
                  "kms:CreateGrant",
                  "kms:Decrypt",
                  "kms:DescribeKey",
                  "kms:Encrypt",
                  "kms:GenerateDataKey",
                  "kms:GenerateDataKeyPair",
                  "kms:GenerateDataKeyPairWithoutPlaintext",
                  "kms:GenerateDataKeyWithoutPlaintext",
                  "kms:ReEncryptTo",
                  "kms:ReEncryptFrom"
              ],
              resources: [`arn:aws:kms:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:key/*`],
              conditions: {"ForAnyValue:StringLike": {"kms:ResourceAliases": `alias/${this.resourcePrefix}-*-key`}},
              }
            ),
            new iam.PolicyStatement({
              effect: iam.Effect.ALLOW,
              actions: ["s3:ListBucket"],
              resources: [
                `arn:aws:s3:::${this.resourcePrefix}-${this.environmentId}-${cdk.Aws.REGION}-${cdk.Aws.ACCOUNT_ID}-*`
              ],
            }),
            new iam.PolicyStatement({
              effect: iam.Effect.ALLOW,
              actions: [
                  "s3:GetObject",
                  "s3:PutObject",
                  "s3:DeleteObject",
              ],
              resources: [
                  `arn:aws:s3:::${this.resourcePrefix}-${this.environmentId}-${cdk.Aws.REGION}-${cdk.Aws.ACCOUNT_ID}-*`
              ],
            }),
            new iam.PolicyStatement({
              effect: iam.Effect.ALLOW,
              actions: [
                  "lakeformation:DeregisterResource",
                  "lakeformation:GetDataAccess",
                  "lakeformation:GrantPermissions",
                  "lakeformation:PutDataLakeSettings",
                  "lakeformation:GetDataLakeSettings",
                  "lakeformation:RegisterResource",
                  "lakeformation:RevokePermissions",
                  "lakeformation:UpdateResource",
                  "glue:CreateDatabase",
                  "glue:CreateJob",
                  "glue:CreateSecurityConfiguration",
                  "glue:DeleteDatabase",
                  "glue:DeleteJob",
                  "glue:DeleteSecurityConfiguration",
                  "glue:GetDatabase",
                  "glue:GetDatabases",
                  "glue:GetMapping",
                  "glue:GetPartition",
                  "glue:GetPartitions",
                  "glue:GetPartitionIndexes",
                  "glue:GetSchema",
                  "glue:GetSchemaByDefinition",
                  "glue:GetSchemaVersion",
                  "glue:GetSchemaVersionsDif",
                  "glue:GetTable",
                  "glue:GetTables",
                  "glue:GetTableVersion",
                  "glue:GetTableVersions",
                  "glue:GetTags",
                  "glue:PutDataCatalogEncryptionSettings",
                  "glue:SearchTables",
                  "glue:TagResource",
                  "glue:UntagResource",
                  "glue:UpdateDatabase",
                  "glue:UpdateJob",
                  "glue:ListSchemas",
                  "glue:ListSchemaVersions"
              ],
              resources: ["*"],
            }),
            new iam.PolicyStatement({
              effect: iam.Effect.ALLOW,
              actions: ["dynamodb:GetItem"],
              resources: [
                `arn:aws:dynamodb:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:table/`,
                `${this.resourcePrefix}-${this.environmentId}-*`,
                `arn:aws:dynamodb:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:table/octagon-*`,
              ],
            }),
          ]
        })
      }
    )
    return glueRole
  }

}
